internal/core/src/common/Array.h

```diff
@@ -0,0 +1,429 @@
+                case DataType::INT64:
+                    return static_cast<T>(
+                        reinterpret_cast<int64_t*>(data_)[index]);
+                case DataType::FLOAT:
+                    return static_cast<T>(
+                        reinterpret_cast<float*>(data_)[index]);
+                case DataType::DOUBLE:
+                    return static_cast<T>(
+                        reinterpret_cast<double*>(data_)[index]);
+                default:
+                    std::cout << datatype_name(element_type_) << std::endl;
+                    PanicInfo("unsupported element type for array");
+            }
+        }
+        return reinterpret_cast<T*>(data_)[index];
+    }
+
+    const std::vector<uint64_t>&
+    get_offsets() const {
+        return offsets_;
+    }
+
+    ScalarArray
+    output_data() const {
+        ScalarArray data_array;
+        switch (element_type_) {
+            case DataType::BOOL: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<bool>(j);
+                    data_array.mutable_bool_data()->add_data(element);
+                }
+                break;
+            }
+            case DataType::INT8:
+            case DataType::INT16:
+            case DataType::INT32: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<int>(j);
+                    data_array.mutable_int_data()->add_data(element);
+                }
+                break;
+            }
+            case DataType::INT64: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<int64_t>(j);
+                    data_array.mutable_long_data()->add_data(element);
+                }
+                break;
+            }
+            case DataType::STRING:
+            case DataType::VARCHAR: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<std::string>(j);
+                    data_array.mutable_string_data()->add_data(element);
+                }
+                break;
+            }
+            case DataType::FLOAT: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<float>(j);
+                    data_array.mutable_float_data()->add_data(element);
+                }
+                break;
+            }
+            case DataType::DOUBLE: {
+                for (int j = 0; j < length_; ++j) {
+                    auto element = get_data<double>(j);
+                    data_array.mutable_double_data()->add_data(element);
+                }
+                break;
+            }
+            default: {
+                PanicInfo("unsupported data type");
+            }
+        }
+        return data_array;
+    }
+
+    int
+    length() const {
+        return length_;
+    }
+
+    int
+    size() const {
```

Comment:
a style suggestion: the return type of `size()` should be `size_t`, the other problem is, `size()` is usually used as the length of container (like `std::vector`, `std::map` do), so this may confuse someone, `ByteSize()` may be a good choice